ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,429 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Máté Kocsis <kocsismate@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_uriparser.h"
+#include "php_uri_common.h"
+#include "Zend/zend_smart_str.h"
+#include "Zend/zend_exceptions.h"
+
+static void uriparser_free_uri(void *uri);
+static void throw_invalid_uri_exception(void);
+
+static inline size_t get_text_range_length(const UriTextRangeA *range)
+{
+	return range->afterLast - range->first;
+}
+
+static void uriparser_copy_uri(UriUriA *new_uriparser_uri, UriUriA *uriparser_uri)
+{
+	ZEND_ASSERT(new_uriparser_uri != NULL && uriparser_uri != NULL);
+
+	int result = uriCopyUriA(new_uriparser_uri, uriparser_uri);
+	ZEND_ASSERT(result == URI_SUCCESS && new_uriparser_uri != NULL);
+}
+
+static void uriparser_normalize_uri(UriUriA *uriparser_uri)
+{
+	int result = uriNormalizeSyntaxExA(uriparser_uri, (unsigned int)-1);
+	ZEND_ASSERT(result == URI_SUCCESS);
+}
+
+static void normalize_uri_if_needed(uriparser_uris_t *uriparser_uris)
```

Comment:
It might be a little nicer to make this:

```
static UriUriA *get_normalized_uri(uriparser_uris_t *uriparser_uris) {
    if (!uriparser_uris->normalized_uri_initialized) {
        // normalize
    }

    return &uriparser_uris->normalized_uri;
}
```

Then you can just is this as a getter for the normalized URI and it will to the right thing.